### PR TITLE
Custom os.path.join that is aware of TF filesystems

### PR DESCRIPTION
**Pull request by adriangb**
Possibly needed for #48086

### CODE REVIEW

**Comment by adriangb** [comment]
@bhack  can you kick off tests here for me? For now just making sure this doesn't break anything

File: tensorflow/python/lib/io/file_io.py
Comment (vnvo2409) [inline tensorflow/python/lib/io/file_io.py:798]: I think it's might be more suitable to check `not (path.startswith("://") or path.startswith("file://"))` because we have a lot of cloud filesystem, like `s3`, `hdfs`, etc. WDYT ?

Comment (adriangb) [reply tensorflow/python/lib/io/file_io.py:798]: I was hoping someone would come along with a better idea :)

It seems like `ram://`, `gc://` or anything else with `://` would not be valid folder or filename is mac, linux or windows (based on [SO](https://stackoverflow.com/a/31976060/6582418)). So maybe we can do:

```python
if `://` in path and not path.startswith("file://")
```

That avoids more complex alternatives like a regex or something.

**Review by bhack — APPROVED** [review]
Just to give a Kokoro pass

**Comment by bhack** [comment]
We could see if this will solve https://github.com/tensorflow/tensorflow/issues/36453

**Comment by adriangb** [comment]
It looks like some sort of internal review is required. Could one of you look into that please? Thanks

**Comment by adriangb** [comment]
@bhack  any idea what this failure means? https://source.cloud.google.com/results/invocations/c285665c-289b-4bef-b8ca-3f87efb37c5f/targets/%2F%2Ftensorflow%2Ftools%2Fapi%2Ftests:api_compatibility_test/tests;group=__main__.ApiCompatibilityTest;test=testAPIBackwardsCompatibility;row=1

The sanity/CPU failures I looked at, I'm waiting to push a fix until all testing finishes

**Comment by bhack** [comment]
If you change the API you need to run https://github.com/tensorflow/tensorflow/issues/44485#issuecomment-787589186

**Comment by adriangb** [comment]
So I'm supposed to run:

```
bazel run tensorflow/tools/api/tests:api_compatibility_test -- --update_goldens True
```

And commit whatever changes that makes?

**Comment by bhack** [comment]
Yes double check the diff but generally are related to your changes.

**Comment by adriangb** [comment]
yeah I imagine this is because I am adding `tf.io.gfile.join`.

**Comment by adriangb** [comment]
@bhack  can you kick off another CI run? I think at least the API and sanity issues should be resolved

**Comment by adriangb** [comment]
And in the meantime: who do we need for API change approval? Might as well get that ball rolling

**Comment by bhack** [comment]
For the API review practices see https://github.com/tensorflow/tensorflow/pull/48161

**Comment by adriangb** [comment]
@bhack another request for a CI run please. Thank you.

File: tensorflow/python/lib/io/file_io.py
Comment (adriangb) [inline tensorflow/python/lib/io/file_io.py:819]: I'm wondering if this logic should be moved into the new `join` method?

**Review by bhack — APPROVED** [review]
Just to trigger Kokoro. Please don't merge this.

**Comment by adriangb** [comment]
Thank you. I fixed the 2 sanity linting issues.

There's no link to see the Windows GPU results, I have no idea why that is failing.

For the plain "Windows" results, I don't see it running `ram_file_system_test`. I mirrored your build changes in #48124 to get it to run.

**Comment by adriangb** [comment]
@tensorflow/api-owners could someone please review this? It mitigates #48086  and unblocks #39609

**Comment by bhack** [comment]
I see the github hints/autocomplete for /cc @tensorflow/api-owners. The tag is used by a reviewer that currently is also a TF member until we will have an official community reviewer status/team (/cc @theadactyl).

**Comment by bhack** [comment]
> For the plain "Windows" results, I don't see it running ram_file_system_test. I mirrored your build changes in #48124 to get it to run.

It is not only the build change you need to move the test itself as you can see in my PR.

**Comment by adriangb** [comment]
Oh I thought I did that :facepalm:

**Comment by adriangb** [comment]
Actually @gbaned it looks like I _did_ move it in ee7a51028fb04dae61936e702449d493997741e8. Can you kick off CI again so we can see if it's working? Thanks!

**Comment by adriangb** [comment]
I can't see the results for `Windows Bazel GPU` (there is no `Details` link, at least not for non-googlers) so I don't know what (if anything) is going wrong there. But other than that all tests seem to be passing, including the moved `ram_filesystem_test` under `Windows Bazel`.

**Review by mihaimaruseac — COMMENTED** [review]
This LGTMs for me but let's wait for discussion on #48124 to see what the best path forward would be

**Comment by bhack** [comment]
We could be ok with this solution  but  I don't know if passing only current test It is enought. 
We could be exposed on other `os.path.join` in the source code that currently have no specific test with the ramfs. Right? 
Need we to search this string in the whole source code?

**Comment by adriangb** [comment]
> We could be exposed on other os.path.join in the source code that currently have no specific test with the ramfs. Right?

I'm not sure I understood. You are saying that `os.path.join` is used somewhere, and in that usage it interacts with the ramfs?

> Need we to search this string in the whole source code?

That is what I did. I searched all of the source for `os.path.join` and replaced it.

**Comment by bhack** [comment]
I meant that many `os.path.join` cases could not have a specific test case with ramfs but if you have already extensively searched all the `os.path.join` occurrences it could be enough. 
We need also to advise to not introduce `os.path.join` in many cases in future commits.

**Comment by bhack** [comment]
Just a note, with this we are not strictly aligned with c++ cause we don't have access to filesystem own separator https://github.com/tensorflow/tensorflow/blob/773a5e411689b04437eb6f431dcdbcbbcec49297/tensorflow/core/platform/file_system.h#L399

In the Python wrapper https://github.com/samikama/tensorflow/blob/master/tensorflow/python/lib/io/file_io_wrapper.cc

**Comment by adriangb** [comment]
This assumes that for native/disk filesystems, `os.path.join` "knows" the right separator to use and that all other filesystems (`ram`, `gc`) `/` is the right separator. I am not sure when this would not be true (which would mean we should reference the C++ defined separator, like you mention above)?

And _if_ a filesystem was introduced that required this, it would just be a question of (1) exposing that file separator to Python and (2) modifying this custom `join` to account for that. With this in place, this has 0 user facing changes, and would not require any edits beyond the `join` method. Without this in place, well you would basically have to make this PR just for that.

**Comment by bhack** [comment]
> This assumes that for native/disk filesystems, os.path.join "knows" the right separator to use and that all other filesystems (ram, gc) / is the right separator. I am not sure when this would not be true (which would mean we should reference the C++ defined separator, like you mention above)?

I meant only that people that are working on c++ need to be aware that every new "://" fs that ovveride the parent separator will need python specific changes or it will be ignored. 

But If we are not going to introduce other "://" fs with its own separator in c++ it is not strictly a problem.

**Comment by adriangb** [comment]
> people that are working on c++ need to be aware that every new "://" fs that ovveride the parent separator will need python specific changes or it will be ignored

Isn't this the current status quo though? It's not like `os.path.join` knows about TFs C++ path seps.

> But If we are not going to introduce other "://" fs with its own separator in c++ it is not strictly a problem.

Yep. But at least know the required Python changes will be centralized in a single place and can be done without any impact on end users.

**Comment by adriangb** [comment]
Based on https://github.com/tensorflow/tensorflow/pull/39609#issuecomment-818999615 it sounds like the plan is to move forward with this PR. Please let me know if there's anything else needed here.

**Comment by adriangb** [comment]
@mihaimaruseac is there anything I can do to move this forward and unblock #39609

**Comment by adriangb** [comment]
There's already some merge conflicts showing up, and given the number of random places this is introducing changes into, I'd expect that they will snowball until this branch is unusable. I'll resolve these conflicts, but I'd appreciate a timely review of the changes.

**Comment by bhack** [comment]
@adriangb I think there is a little bit of overhead with the 2.5 release, so you need to wait a little bit for the review.

**Comment by adriangb** [comment]
Understood, I'll try to keep resolving conflicts for now.

File: tensorflow/core/platform/BUILD
Comment (mihaimaruseac) [inline tensorflow/core/platform/BUILD:431]: Should we also delete/move this file?

Comment (adriangb) [reply tensorflow/core/platform/BUILD:431]: Sorry if I am misunderstanding, but I believe `ram_file_system_test.py` _is_ being moved: https://github.com/tensorflow/tensorflow/commit/ee7a51028fb04dae61936e702449d493997741e8#diff-18772cb1894d1c5e6364f679cee08b916774e945094644a52b17735188b9f907

Comment (mihaimaruseac) [reply tensorflow/core/platform/BUILD:431]: Sorry, I had the page displaying the status wrong.

**Review by mihaimaruseac — APPROVED** [review]

**Comment by adriangb** [comment]
I looked at the CI failures, I think none of them are related to this PR, please correct me if I'm wrong and I will hunt down the issue.

File: tensorflow/python/lib/io/file_io.py
Comment (mihaimaruseac) [inline tensorflow/python/lib/io/file_io.py:782]: > We should add tests for this new API endpoint in file_io_test.py

(from internal review)

File: tensorflow/python/lib/io/file_io.py
Comment (mihaimaruseac) [inline tensorflow/python/lib/io/file_io.py:789]: > Maybe add examples here of the TF filesystem behavior?

(from internal review)

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

File: tensorflow/python/lib/io/file_io.py
Comment (adriangb) [reply tensorflow/python/lib/io/file_io.py:782]: That's a good idea.
I added some tests in `file_io_test.py`, very similar to what was already in `ram_file_system_test.py ` but also testing `gcs://` and that the behavior for native filesystems.

File: tensorflow/python/lib/io/file_io.py
Comment (adriangb) [reply tensorflow/python/lib/io/file_io.py:797]: I added an example in the docstring.

File: tensorflow/python/lib/io/file_io_test.py
Comment (adriangb) [inline tensorflow/python/lib/io/file_io_test.py:89]: @mihaimaruseac let me know if this is the desired behavior. I realize that `file://` is the native filesystem, but maybe it should be treated like other ulr-like filesystems? What does `GFile(join("file://", "dir", "file.py"))` expect?

**Review by adriangb — COMMENTED** [review]
Thank you for the review comments @mihaimaruseac . I am leaving a followup regarding the intended behavior of `join` with `file://`.

File: tensorflow/python/lib/io/file_io_test.py
Comment (mihaimaruseac) [reply tensorflow/python/lib/io/file_io_test.py:89]: I think it should be `"file://dir/file.py"`. It's similar to URLs

**Review by mihaimaruseac — APPROVED** [review]

File: tensorflow/python/lib/io/file_io_test.py
Comment (adriangb) [reply tensorflow/python/lib/io/file_io_test.py:89]: easy enough change; and it simplifies the implementation/testing.

File: tensorflow/python/lib/io/file_io.py
Comment (rohan100jain) [reply tensorflow/python/lib/io/file_io.py:797]: Could you add a slightly more complex example here? I tried the provided example with os.path.join and I see the same result. So might be helpful to find something that has a different result....

Comment (adriangb) [reply tensorflow/python/lib/io/file_io.py:797]: results would only differ from os.path.join on Windows

Comment (adriangb) [reply tensorflow/python/lib/io/file_io.py:797]: On windows you get:

```
>>> tf.io.join("gcs://folder", "file.py") 
"gcs://folder\\file.py"
```

Comment (adriangb) [reply tensorflow/python/lib/io/file_io.py:798]: I updated the docstring to:
```
This is the same as os.path.join except that it guarantees correct
handling of tensorflow specific filesystems like `gcs://` and `ram://`
on all platforms.
```
To try to clarify. Let me know if that works or if there's any other wording you'd prefer.

File: tensorflow/python/lib/io/file_io_test.py
Comment (adriangb) [reply tensorflow/python/lib/io/file_io_test.py:89]: @mihaimaruseac I just pushed this change, please re-approve so CI runs when you get a chance 😃

**Review by mihaimaruseac — APPROVED** [review]

**Review by bhack — APPROVED** [review]

**Comment by adriangb** [comment]
@bhack it looks like [last run only a single doctest target was failing](https://source.cloud.google.com/results/invocations/ce4b6da3-7984-4055-a504-a0c546c50b12/targets). I attempted to fix it. Could you approve so CI runs again? Thank you

**Review by bhack — APPROVED** [review]

**Review by mihaimaruseac — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Comment by bhack** [comment]
We still have a doctest issue.

**Comment by adriangb** [comment]
> We still have a doctest issue.

Yup, sorry for taking a bit to work on it, I was vacationing/traveling the last couple of weeks. I'll take another stab at it today, I might need to simplify the multi-statement docstring.

If you have any suggestions for something simpler that might still get the point across, I'm all ears.

**Comment by adriangb** [comment]
@bhack I think I fixed the syntax issue. I can't really test locally (requires compiling TF), so could you please kick off CI again to check? Thanks

**Review by bhack — APPROVED** [review]

**Comment by bhack** [comment]
Doctest Is failing again.

**Comment by adriangb** [comment]
Ah yeah I see the issue, should be fixed now (I tested with vanilla doctest and w/o the TF specific stuff and it works). Sorry for the back and forth.

**Comment by bhack** [comment]
Yes we are increasing the carbon footprint with this commits/ping/CI loops (as we rebuild all every time). :grimacing: 

I hope we could find soon a more accessible local testing solution for contributors /cc @mihaimaruseac @theadactyl

**Review by bhack — APPROVED** [review]
invoke Kokoro

**Comment by adriangb** [comment]
Yeah, in this specific case, I realized I could test most of the issues by throwing the test into a `test.py` and running stdlib doctest on it (i.e. I was just being lazy / not coming up with alternatives).

But in general, that's not possible (eg. with #39609 there's no way to test it). I do see all of the work you're all putting into making it easier, very much appreciated! FWIW I think the [move of Keras to its own repo](https://twitter.com/fchollet/status/1404967230048149506) will help a ton.

File: tensorflow/python/saved_model/BUILD
Comment (adriangb) [inline tensorflow/python/saved_model/BUILD:412]: Looks like this needs to be sorted (moved one down). I'll push it once current CI finishes.

File: tensorflow/python/platform/ram_file_system_test.py
Comment (adriangb) [inline tensorflow/python/platform/ram_file_system_test.py:175]: Since I added `testJoinUrlLike` and `testJoinFilesystem` in `file_io_test.py`, should we minimize the changes by not adding this test (which would also mean not moving the test file)?

**Comment by bhack** [comment]
> But in general, that's not possible (eg. with #39609 there's no way to test it). I do see all of the work you're all putting into making it easier, very much appreciated! 

You can still theoretically execute a test subset with something like https://github.com/tensorflow/build/issues/23 as it is python only.

> FWIW I think the move of Keras to its own repo will help a ton.

In Keras only PRs yes but In cases like this one it will be more complex as you are touching Keras and non Keras files in a single PR right?

So you will need to have two separate review "in a dependency flow" and someone that has the full vision on what you are contributing and so why you are introducing a change in TF to feed something in Keras. See also [this thread in the forum](https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999/4?u=bhack).

**Comment by adriangb** [comment]
Yeah, there are definitely other complexities that will arise from that move. I'll check out the links, thanks for those!

**Review by mihaimaruseac — APPROVED** [review]

**Review by mihaimaruseac — APPROVED** [review]

**Comment by adriangb** [comment]
@mihaimaruseac looks like the `MacOS CPU Python 3` and `Windows Bazel GPU` failures are the same as on `master` (i.e. unrelated to this PR), so everything here is working. Could you chime in on https://github.com/tensorflow/tensorflow/pull/48125#discussion_r654591774? Thanks

File: tensorflow/python/platform/ram_file_system_test.py
Comment (mihaimaruseac) [reply tensorflow/python/platform/ram_file_system_test.py:175]: Sounds good to me

Comment (adriangb) [reply tensorflow/python/platform/ram_file_system_test.py]: Cool, reverted those changes. Would you mind kicking off CI? Thanks

**Review by bhack — APPROVED** [review]

**Review by mihaimaruseac — APPROVED** [review]

**Comment by adriangb** [comment]
@mihaimaruseac I don't think any of the failures are related to these changes, please correct me if this is not the case and something needs to be fixed, or if you want me to rebase and re-run with the latest master changes.

**Comment by bhack** [comment]
Copybara import is failing. We need to wait for an internal feedback.

File: tensorflow/python/lib/io/file_io.py
Comment (mihaimaruseac) [inline tensorflow/python/lib/io/file_io.py:812]: > Docstring contains a backslash, use r""" instead of """. [g-docstring-has-escape]

Please fix.

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]
We'll have to split the keras parts to the Keras repo now that it got split.

**Comment by adriangb** [comment]
> We'll have to split the keras parts to the Keras repo now that it got split.

So just revert any changes to `tensorflow/python/keras/*` files?

File: tensorflow/python/lib/io/file_io.py
Comment (adriangb) [reply tensorflow/python/lib/io/file_io.py:812]: attempted!

**Review by gbaned — APPROVED** [review]

**Comment by mihaimaruseac** [comment]
> > We'll have to split the keras parts to the Keras repo now that it got split.
> 
> So just revert any changes to `tensorflow/python/keras/*` files?

Yes, they'll have to move to the Keras repo isntead

**Comment by adriangb** [comment]
Ok, reverted all Keras changes.

**Review by bhack — APPROVED** [review]
For kokoro

**Review by mihaimaruseac — APPROVED** [review]

**Comment by adriangb** [comment]
@mihaimaruseac failures looked like build / compilation failures to me, guessing they're unrelated to these changes. I merged in master to see if it helps.

**Review by mihaimaruseac — APPROVED** [review]

**Comment by adriangb** [comment]
@mihaimaruseac only failure is Windows Bazel GPU, and the logs for that aren't public so I can't diagnose issues. Let me know if there's anything you need from me. Thanks.

**Comment by mihaimaruseac** [comment]
The issue is some internal builds are failing and will need some time to look into those

**Pull request merged**
